src/collection/historical_signatures/signature_collector.py

```diff
@@ -0,0 +1,215 @@
+        """
+        Flag to mark completion of collection.
+        """
+        return self._signatures_completed
+
+    @log_performance_time(LOGGER)
+    def _get_assignment(self) -> None:
+        """
+        Obtain assignment for data collection.
+        """
+        # if oldest obtained signature is defined, proceed to collection
+        if self._oldest_signature:
+            return
+        # otherwise, get retrieve oldest signature from database.
+        with db.get_db_session() as session:
+            LOGGER.info("Getting the oldest stored signature for protocol = {}.".format(self.protocol))
+            # Get the last signature collector recorded by `SignatureCollector` for given protocol
+            oldest_signature = session.query(db.TransactionStatusWithSignature.signature) \
+                .filter(db.TransactionStatusWithSignature.source == self.protocol) \
+                .filter(db.TransactionStatusWithSignature.collection_stream == db.CollectionStreamTypes.SIGNATURE) \
+                .order_by(db.TransactionStatusWithSignature.id.desc()) \
+                .first()
+            # If no signatures collected yet, get signature from watershed block
+            if not oldest_signature:
+                LOGGER.warning(
+                    "No signatures found for protocol = {}.".format(self.protocol)
+                )
+                self._get_watershed_block_signature()
+                return
+
+            oldest_signature = oldest_signature.signature
+            LOGGER.info("The oldest stored signature = {} for protocol = {}.".format(oldest_signature, self.protocol))
+            self._oldest_signature = Signature.from_string(str(oldest_signature))
+
+    @log_performance_time(LOGGER)
+    def _get_data(self) -> None:
+        """
+        Collect signatures with metadata through Solana API.
+        """
+        try:
+            self._fetch_signatures()
+        # catch any possible API's hallucinations, try once again
+        except Exception as e:  # pylint: disable=broad-exception-caught
+            tb_str = traceback.format_exc()
+            # Log the error message along with the traceback
+            LOGGER.error(f"While fetching signatures for `{self.protocol}` before `{self._oldest_signature}`"
+                         f" error occurred: {e}\nTraceback:\n{tb_str}")
+            self._fetch_signatures()
+
+        LOGGER.info(f"Collected {len(self._collected_signatures)} signatures for `{self.protocol}")
+
+    @log_performance_time(LOGGER)
+    def _write_tx_data(self) -> None:
+        """
+        Write signatures and tx meta data to database.
+        """
+        with db.get_db_session() as session:
+            if not self._collected_signatures:
+                return
+            for signature in self._collected_signatures:
+                # store only transactions from slot with complete transaction history fetched.
+                if signature.slot < self._oldest_completed_slot:
```

Comment:
How can this happen? :thinking:

Reply:
```
b3(s7 s6) b2(s5 s4) b1( s3 s2) b0(s1 s0)
```
if in the first itteration we collect signatures from s6 to s1, we only store signatures from blocks b2 and b1. then in next iter. we start collecting before presumably the last signature in b1, therefore we collect s2, s1, s0, s(-1).  No need to store s1 in the first iterration.